tests/unix_datagram.rs

```diff
@@ -0,0 +1,436 @@
+    assert_ok!(datagram2.connect(&path1));
+    assert_ok!(poll
+        .registry()
+        .reregister(&datagram1, TOKEN_1, Interests::WRITABLE));
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![ExpectEvent::new(TOKEN_1, Interests::WRITABLE)],
+    );
+}
+
+#[test]
+fn unix_datagram_deregister() {
+    let (mut poll, mut events) = init_with_poll();
+    let dir = assert_ok!(TempDir::new("unix_datagram"));
+    let path1 = dir.path().join("one");
+    let path2 = dir.path().join("two");
+
+    let datagram1 = assert_ok!(UnixDatagram::bind(&path1));
+    assert_ok!(poll
+        .registry()
+        .register(&datagram1, TOKEN_1, Interests::WRITABLE));
+
+    let datagram2 = assert_ok!(UnixDatagram::bind(&path2));
+    assert_ok!(datagram2.connect(&path1));
+    assert_ok!(poll.registry().deregister(&datagram1));
+    expect_no_events(&mut poll, &mut events);
+}
+
+fn smoke_test_unconnected(datagram1: UnixDatagram, datagram2: UnixDatagram) {
+    let (mut poll, mut events) = init_with_poll();
+
+    let addr1 = assert_ok!(datagram1.local_addr());
+    let addr2 = assert_ok!(datagram2.local_addr());
+    let path1 = addr1.as_pathname().expect("failed to get pathname");
+    let path2 = addr2.as_pathname().expect("failed to get pathname");
+
+    assert_ok!(poll.registry().register(
+        &datagram1,
+        TOKEN_1,
+        Interests::READABLE.add(Interests::WRITABLE)
+    ));
+    assert_ok!(poll.registry().register(
+        &datagram2,
+        TOKEN_2,
+        Interests::READABLE.add(Interests::WRITABLE)
+    ));
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![
+            ExpectEvent::new(TOKEN_1, Interests::WRITABLE),
+            ExpectEvent::new(TOKEN_2, Interests::WRITABLE),
+        ],
+    );
+
+    let mut buf = [0; DEFAULT_BUF_SIZE];
+    assert_would_block(datagram1.recv_from(&mut buf));
+
+    assert_ok!(datagram1.send_to(DATA1, path2));
+    assert_ok!(datagram2.send_to(DATA2, path1));
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![
+            ExpectEvent::new(TOKEN_1, Interests::READABLE),
+            ExpectEvent::new(TOKEN_2, Interests::READABLE),
+        ],
+    );
+
+    let (read, from_addr1) = assert_ok!(datagram1.recv_from(&mut buf));
+    assert_eq!(read, DATA2.len());
+    assert_eq!(buf[..read], DATA2[..]);
+    assert_eq!(
+        from_addr1.as_pathname().expect("failed to get pathname"),
+        path2
+    );
+
+    let (read, from_addr2) = assert_ok!(datagram2.recv_from(&mut buf));
+    assert_eq!(read, DATA1.len());
+    assert_eq!(buf[..read], DATA1[..]);
+    assert_eq!(
+        from_addr2.as_pathname().expect("failed to get pathname"),
+        path1
+    );
+
+    assert!(assert_ok!(datagram1.take_error()).is_none());
+    assert!(assert_ok!(datagram2.take_error()).is_none());
+}
+
+fn smoke_test_connected(datagram1: UnixDatagram, datagram2: UnixDatagram) {
+    let (mut poll, mut events) = init_with_poll();
+
```

Comment:
Maybe add a check that they are bound connected (to each other), via `{local,peer}_addr`?